zebra-consensus/src/block.rs

```diff
@@ -77,9 +77,24 @@ where
 
         // TODO(jlusby): Error = Report, handle errors from state_service.
         async move {
+            let hash = block.hash();
+
+            // These checks only apply to generated blocks. We check the block
+            // height for parsed blocks when we deserialize them.
+            let height = block
+                .coinbase_height()
+                .ok_or("Invalid block: missing block height.")?;
+            if height > BlockHeight::MAX {
+                Err("Invalid block height: greater than the maximum height.")?;
+            }
+
+            // Check that this block is actually a new block
+            if BlockVerifier::get_block(&mut state, hash).await?.is_some() {
+                Err(format!("Block has already been verified. {:?} {:?}", height, hash))?;
+            }
```

Comment:
Why do we want to deduplicate block verifications here?

Reply:
There are a few reasons:
* The Sync service sometimes downloads the same block multiple times
* It's a waste of CPU to verify a block multiple times
* It's useful to know the height and hash of duplicate blocks

But I think this change might overlap with a similar change in the Sync service:
https://github.com/ZcashFoundation/zebra/pull/853/files#diff-d07923e0d6c7a0cac7a37fc2642abef5R397

(It's not completely redundant, because there is still a race between multiple download and verify tasks. But it won't happen very often.)

Reply:
This seems like a race condition waiting to happen (because it's reported as a failure, which is wrong if the other verification was a success).

Reply:
Okay, cool.  Is it accurate that the first reason is the primary one?  While the sync service is still buggy I think this makes sense, but I wonder whether -- if that were fixed -- the overhead of the state service call is worth it.  It looks like the current implementation of `get_block` will pull the entire block out of the state.  Maybe we could just fetch the depth, or accept the possibility of duplicating verification work.  We could probably identify the hash and height of duplicate blocks in the state service instead of here (at the point that someone requests to commit an existing block).

Reply:
> This seems like a race condition waiting to happen (because it's reported as a failure, which is wrong if the other verification was a success).

Hmm, I think you're right, we could log at info and return success instead?

Reply:
> Okay, cool. Is it accurate that the first reason is the primary one? While the sync service is still buggy I think this makes sense, but I wonder whether -- if that were fixed -- the overhead of the state service call is worth it.

I don't think the overhead is worth it, there's a bunch of stuff I've added as temporary sync diagnostics. I didn't put a comment on this one though.

> It looks like the current implementation of `get_block` will pull the entire block out of the state. Maybe we could just fetch the depth, or accept the possibility of duplicating verification work. We could probably identify the hash and height of duplicate blocks in the state service instead of here (at the point that someone requests to commit an existing block).

I think we should move all the duplicate block checks to the state, log an info-level message, and turn the errors into successes.

Reply:
See #865 for the cleanup of the errors and state checks.